src/decoder_fluidsynth.cpp

```diff
@@ -0,0 +1,241 @@
+}
+
+static int vio_read(void *ptr, int count, void* userdata) {
+	auto* f = reinterpret_cast<Filesystem_Stream::InputStream*>(userdata);
+	return f->read(reinterpret_cast<char*>(ptr), count).gcount();
+}
+
+static int vio_seek(void* userdata, long offset, int origin) {
+	auto* f = reinterpret_cast<Filesystem_Stream::InputStream*>(userdata);
+	if (f->eof()) f->clear(); // emulate behaviour of fseek
+
+	f->seekg(offset, Filesystem_Stream::CSeekdirToCppSeekdir(origin));
+
+	return f->tellg();
+}
+
+static long vio_tell(void* userdata) {
+	auto* f = reinterpret_cast<Filesystem_Stream::InputStream*>(userdata);
+	return f->tellg();
+}
+
+static int vio_close(void* userdata) {
+	auto stream_ref = reinterpret_cast<Filesystem_Stream::InputStream*>(userdata);
+	delete stream_ref;
+
+	return 0;
+}
+
+#ifdef HAVE_FLUIDLITE
+static fluid_fileapi_t fluidlite_vio = {
+	nullptr,
+	nullptr,
+	vio_open,
+	vio_read,
+	vio_seek,
+	vio_close,
+	vio_tell
+};
+#endif
+
+struct FluidSettingsDeleter {
+	void operator()(fluid_settings_t* s) const {
+		delete_fluid_settings(s);
+	}
+};
+
+struct FluidSynthDeleter {
+	void operator()(fluid_synth_t* s) const {
+		delete_fluid_synth(s);
+	}
+};
+
+namespace {
+	std::unique_ptr<fluid_settings_t, FluidSettingsDeleter> global_settings;
+	std::unique_ptr<fluid_synth_t, FluidSynthDeleter> global_synth;
+#ifdef HAVE_FLUIDSYNTH
+	fluid_sfloader_t* global_loader; // owned by global_settings
+#endif
+	int instances = 0;
+}
+
+static fluid_synth_t* create_synth(std::string& error_message) {
+	fluid_synth_t* syn = new_fluid_synth(global_settings.get());
+
+#ifdef HAVE_FLUIDSYNTH
+	fluid_synth_add_sfloader(syn, global_loader);
+#endif
+
+	if (fluid_synth_sfload(syn, "easyrpg.soundfont", 1) == FLUID_FAILED) {
+		error_message = "Could not load soundfont.";
+		return nullptr;
+	}
+
+	fluid_synth_set_interp_method(syn, -1, 7);
+
+	return syn;
+}
+
+FluidSynthDecoder::FluidSynthDecoder() {
+	++instances;
+
+	// Optimisation: Only create the soundfont once and share the synth
+	// Sharing is only not possible when a Midi is played as a SE (unlikely)
+	if (instances > 1) {
+		std::string error_message;
+		instance_synth = create_synth(error_message);
+		if (!instance_synth) {
+			// unlikely, the SF was already allocated once
+			Output::Debug("FluidSynth failed: {}", error_message);
+		}
+	} else {
+		instance_synth = global_synth.get();
+		fluid_synth_program_reset(global_synth.get());
+	}
+}
+
+FluidSynthDecoder::~FluidSynthDecoder() {
+	--instances;
+	assert(instances >= 0);
+
+	if (instance_synth != global_synth.get()) {
+		delete_fluid_synth(instance_synth);
+	}
+}
+
+bool FluidSynthDecoder::Initialize(std::string& error_message) {
+	static bool init = false;
+	static bool once = false;
+
+	// only initialize once
+	if (once)
+		return init;
+	once = true;
+
+#ifdef HAVE_FLUIDLITE
+	fluid_set_default_fileapi(&fluidlite_vio);
+#endif
+
+	if (!global_settings) {
+		global_settings.reset(new_fluid_settings());
```

Comment:
Are chorus and reverberation disabled by default? According to the FluidSynth 2.0 API, they are enabled by default, not just in the command line player. They are a huge CPU hog, even more than high sampling rates.

Tuning buffer size and count also affects cpu/latency significantly, and FluidSynth seems to use a small buffer oriented for pro audio low latency, so CPU usage can be reduced there, too. I didn't check how Fluid buffer is being filled here or if needs to be tweaked to match the expected Player audio buffer size. FluidSynth API has some buffer related settings related to period-size and periods. Maybe these are for its own playback.

On Linux, setting realtime-prio to 0 disables Fluid attempting higher priority, which prevents some permission warnings, likely not requiring guards as should be ignored on other platforms.

See #17 for some early CLI tests.

Reply:
Here are the default settings:
```
  fluid_settings_register_str(settings, "synth.verbose", "no", 0, NULL, NULL);
  fluid_settings_register_str(settings, "synth.dump", "no", 0, NULL, NULL);
  fluid_settings_register_str(settings, "synth.reverb.active", "yes", 0, NULL, NULL);
  fluid_settings_register_str(settings, "synth.chorus.active", "yes", 0, NULL, NULL);
  fluid_settings_register_str(settings, "synth.ladspa.active", "no", 0, NULL, NULL);
  fluid_settings_register_str(settings, "midi.portname", "", 0, NULL, NULL);
  fluid_settings_register_str(settings, "synth.drums-channel.active", "yes", 0, NULL, NULL);

  fluid_settings_register_int(settings, "synth.polyphony",
                             256, 16, 4096, 0, NULL, NULL);
  fluid_settings_register_int(settings, "synth.midi-channels",
                             16, 16, 256, 0, NULL, NULL);
  fluid_settings_register_num(settings, "synth.gain",
                             0.2f, 0.0f, 10.0f,
                             0, NULL, NULL);
  fluid_settings_register_int(settings, "synth.audio-channels",
                             1, 1, 256, 0, NULL, NULL);
  fluid_settings_register_int(settings, "synth.audio-groups",
                             1, 1, 256, 0, NULL, NULL);
  fluid_settings_register_int(settings, "synth.effects-channels",
                             2, 2, 2, 0, NULL, NULL);
  fluid_settings_register_num(settings, "synth.sample-rate",
                             44100.0f, 22050.0f, 96000.0f,
                             0, NULL, NULL);
  fluid_settings_register_int(settings, "synth.min-note-length", 10, 0, 65535, 0, NULL, NULL);
```

About 2nd and 3rd paragraph: The real-time stuff is not applicable because fluidlite does not start any daemon. The buffer/latency is the default buffer size. This means for fast music it will sound out-of-sync because the MIDI resolution is ~4ms. This will be resolved after tyrone-sudeiums PR. Then the MIDI can play in real-time and the system ALSA Midi player can be used.

Reply:
Shall I disable reverb and chorus everywhere or only on slower platforms? Is there a hearable difference?

Reply:
Maybe only on slower platforms? I don't want you destroyed my beautiful game music when I play on my PC :)

Also, once all this is done can one of you guys catalogue all the audio configuration knobs in an github issue? These things should all eventually be modifiable via command line / config / scene settings.

Reply:
Or rather, are these features enabled in RPG_RT? We could follow the standard as our default.

Reply:
Default playback with DirectMusic on RPG_RT seems to use a barely hearable chorus and reverb, I feel them mostly undetectable by most users. Default chorus and reverb thresholds on FluidSynth are too high and sounds pretty different compared to the original RPG_RT. It sounds a lot closer when chorus and reverb are off, actually. I'd suggest hear this by your self with scc1t2.sf2 at 22050 with disabled chorus and reverb to compare.

Reply:
well there could be a config option to turn them on again later, so I make it default OFF for now. Configuration can be decided when the config code is more advanced. I will prepare a list for you with all useful settings per individual audio decoder.

When the system-wide MIDI support is ready this is also a non-issue for Win/Lin/macOS because the OS configuration is used (but don't ask me how to correctly configure MIDI out on Linux xD)